### PR TITLE
Stop folder when running out of disk space (fixes #2057)

### LINKED ISSUE #2057: enhancement: Stop sync when free disk space reaches a minimum level
Feature request:
Instead of continuing to sync files until a device runs out of space and file creation fails, have an option to stop syncing the folder and show an error when the free disk space reaches a user-defined threshold (e.g. 2 Gib).

Knowing a folder's Global State and the amount of files Out of Sync on the local system, it would be possible for Syncthing to tell when a complete sync is going to exceed available space, and warn the user in advance. This would avoid worse problems caused when disk space reaches zero.

(Now I see @Zillode 's comment in #1790 that there is no golang library for getting free disk space. Any news or other thoughts on that matter?) 

Closed issues related to out-of-space errors: 
https://github.com/syncthing/syncthing/issues/302
https://github.com/syncthing/syncthing/issues/1790

### PR DESCRIPTION
**Pull request by Zillode**

### CODE REVIEW

**Comment by calmh** [comment]
To add the new dependency:

```
go run build.go setup
go run build.go deps
```

File: internal/config/config.go [hunk 1]
```diff
@@ -462,6 +466,14 @@ func ChangeRequiresRestart(from, to Configuration) bool {
 	return false
 }
 
+func convertV10V11(cfg *Configuration) {
+	// Set minimum disk space on existing folders to 500MB
+	for i := range cfg.Folders {
+		cfg.Folders[i].MinimumDiskSpaceMB = 500
```
Comment (calmh) [inline internal/config/config.go]: Not sure I like going by MB and defaulting to this. What about when if we sync 10 MB of files into an empty 256 MB flash card on an embedded system - shouldn't that work out of the box?

File: internal/osutil/osutil.go [hunk 2]
```diff
@@ -210,3 +213,8 @@ func init() {
 func IsWindowsExecutable(path string) bool {
 	return execExts[strings.ToLower(filepath.Ext(path))]
 }
+
+func FreeDiskSpaceMB(path string) (uint64, error) {
```
Comment (calmh) [inline internal/osutil/osutil.go]: I suggest `int64` and just leaving it as bytes here. (We don't use unsigned ints for arithmetic, for safety reasons.)

File: internal/model/model.go [hunk 3]
```diff
@@ -1706,6 +1706,8 @@ func (m *Model) CheckFolderHealth(id string) error {
 			err = errors.New("folder path missing")
 		} else if !folder.HasMarker() {
 			err = errors.New("folder marker missing")
+		} else if space, syserr := osutil.FreeDiskSpaceMB(folder.Path()); syserr != nil && space < uint64(folder.MinimumDiskSpaceMB) {
```
Comment (calmh) [inline internal/model/model.go]: We might still fill the disk if we're trying to pull a file that's larger than 500 MB.

File: internal/config/config.go [hunk 4]
```diff
@@ -67,19 +67,20 @@ func (cfg Configuration) Copy() Configuration {
...
-	Order           PullOrder                   `xml:"order" json:"order"`
-	IgnoreDelete    bool                        `xml:"ignoreDelete" json:"ignoreDelete"`
+	ID               string                      `xml:"id,attr" json:"id"`
+	RawPath          string                      `xml:"path,attr" json:"path"`
+	Devices          []FolderDeviceConfiguration `xml:"device" json:"devices"`
+	ReadOnly         bool                        `xml:"ro,attr" json:"readOnly"`
+	RescanIntervalS  int                         `xml:"rescanIntervalS,attr" json:"rescanIntervalS"`
+	IgnorePerms      bool                        `xml:"ignorePerms,attr" json:"ignorePerms"`
+	AutoNormalize    bool                        `xml:"autoNormalize,attr" json:"autoNormalize"`
+	MinimumDiskFreeP int                         `xml:"MinimumDiskFreeP,attr" json:"MinimumDiskFreeP"`
```
Comment (calmh) [inline internal/config/config.go]: I think `MinDiskFreePct` and `minDiskFreePct` as serialization

File: internal/config/wrapper.go [hunk 5]
```diff
@@ -96,6 +96,10 @@ func Load(path string, myID protocol.DeviceID) (*Wrapper, error) {
 	return Wrap(path, cfg), nil
 }
 
+func (w *Wrapper) Path() string {
```
Comment (calmh) [inline internal/config/wrapper.go]: This is the path of the config file? I think you want the `FolderConfiguration.Path()`?

File: internal/model/model.go [hunk 6]
```diff
@@ -1653,6 +1653,14 @@ func (m *Model) BringToFront(folder, file string) {
 // CheckFolderHealth checks the folder for common errors and returns the
 // current folder error, or nil if the folder is healthy.
 func (m *Model) CheckFolderHealth(id string) error {
+	// This could be expensive wrt sleeping disks... if random(1) < 0.2 ?
+	usage, syserr := osutil.DiskUsage(m.cfg.Path())
+	if syserr == nil && usage > 99 {
```
Comment (calmh) [inline internal/model/model.go]: Should use configured value here I think. :)

File: internal/osutil/osutil.go [hunk 7]
```diff
@@ -210,3 +211,13 @@ func init() {
 func IsWindowsExecutable(path string) bool {
 	return execExts[strings.ToLower(filepath.Ext(path))]
 }
+
+func DiskDetails(path string) (int64, int64, error) {
```
Comment (calmh) [inline internal/osutil/osutil.go]: I'm not super excited about this name as it's not clear what it returns exactly. It could be clarified by something like

`func DiskUsage(path string) (size, free int64, err error)`

and then have the other one below be `func DiskUsedPercent(path string) float64` or so.

File: internal/model/rwfolder.go [hunk 8]
```diff
@@ -492,6 +494,18 @@ func (p *rwFolder) pullerIteration(ignores *ignore.Matcher) int {
 		return true
 	})
 
+	// Check if we are able to store all files on disk
+	if changed > 0 {
+		folder, ok := p.model.cfg.Folders()[p.folder]
+		if ok {
+			diskSize, diskFree, diskErr := osutil.DiskDetails(p.folder)
+			if diskErr == nil && (diskSize-diskFree)-pullFileSize < (diskSize/100)*int64(folder.MinimumDiskFreeP) {
```
Comment (calmh) [inline internal/model/rwfolder.go]: Should this simply be `if diskFree < pullFileSize`?

Comment (Zillode) [inline internal/model/rwfolder.go]: Could work, but we could anticipate it sooner with MinDiskFreePct. Your call

**Comment by Zillode** [comment]
Do not merge, wait for upstream patch: https://github.com/ricochet2200/go-disk-usage/pull/1

**Comment by Zillode** [comment]
@st-jenkins run integrations tests on windows please

File: internal/model/rwfolder.go [hunk 9]
```diff
@@ -492,6 +494,18 @@ func (p *rwFolder) pullerIteration(ignores *ignore.Matcher) int {
 		return true
 	})
 
+	// Check if we are able to store all files on disk
+	if pullFileSize > 0 {
+		folder, ok := p.model.cfg.Folders()[p.folder]
+		if ok {
+			diskSize, diskFree, diskErr := osutil.DiskUsage(p.folder)
+			if diskErr == nil && (diskSize-diskFree)-pullFileSize < (diskSize/100)*int64(folder.MinDiskFreePct) {
```
Comment (calmh) [inline internal/model/rwfolder.go]: This is too much arithmetic in the if clause, I find it hard to follow and I'm not sure it's correct. The current amount of used space (size - free), minus the space we need, less than the percentage specified?

I would expect the checkFolderHealth thing to take care of the percentage, and here just check if we have space for the files we are about to pull, i.e. something like diskFree < pullFileSize?

File: internal/model/model.go [hunk 10]
```diff
@@ -1653,6 +1653,14 @@ func (m *Model) BringToFront(folder, file string) {
 // CheckFolderHealth checks the folder for common errors and returns the
 // current folder error, or nil if the folder is healthy.
 func (m *Model) CheckFolderHealth(id string) error {
+	// This could be expensive wrt sleeping disks... if random(1) < 0.2 ?
+	usage, syserr := osutil.DiskUsedPercentage(m.cfg.ConfigPath())
+	if syserr == nil && usage > 99 {
```
Comment (calmh) [inline internal/model/model.go]: syserr is an unusual variable name, would `if usage, err := osutil.DiskUsedPercentage(...); err == nil && usage > theConfigValue {` suffice here?

File: internal/model/model.go [hunk 11]
```diff
@@ -1668,6 +1676,8 @@ func (m *Model) CheckFolderHealth(id string) error {
 			err = errors.New("folder path missing")
 		} else if !folder.HasMarker() {
 			err = errors.New("folder marker missing")
+		} else if usage, syserr := osutil.DiskUsedPercentage(folder.Path()); syserr == nil && usage > float64(100-folder.MinDiskFreePct) {
```
Comment (calmh) [inline internal/model/model.go]: This is fine, apart from s/syserr/err/ I think

File: internal/model/model.go (hunk 10 above)
Comment (Zillode) [inline internal/model/model.go]: 1. I think we should panic here, the db could become corrupted at any time now. It's just that the panic gets recovered somehow. But if you don't like the panic I won't investigate it
2. What about wrapping it with ```if rand.Float64() < 0.2``` to keep things cheap? Should I add it?

**Comment by Zillode** [comment]
Strange, netbsd fails to build? Works on my machine non-dockerized.

**Comment by Zillode** [comment]
@st-jenkins run integration tests on linux and windows please

**Comment by Zillode** [comment]
I think we need a beta channel to properly test this before deploying it to everyone:)
Maybe we can use the strategy of android, namely delivering the new release to the lucky 5%?

**Comment by calmh** [comment]
It needs to actually build though. :) I don't particularly care if the functionality is just stubbed out for netbsd to start with, if it works for the major platforms.

**Comment by Zillode** [comment]
Would it be better if ```calmh/du``` provides an error if the platform is not supported?
Without this we need to list the supported platforms here, which could become out of sync with the upstream package.

**Comment by calmh** [comment]
I had totally forgotten that I had written that package. Yes, I guess it should!

**Comment by calmh** [comment]
@Zillode calmh/du now builds on everything, but always returns an error on netbsd, openbsd and solaris.

File: internal/model/model.go [hunk 12]
```diff
@@ -1658,13 +1662,19 @@ func (m *Model) BringToFront(folder, file string) {
...
+		return errors.New("out of disk space")
+	}
+
 	folder, ok := m.cfg.Folders()[id]
 	if !ok {
 		return errors.New("folder does not exist")
 	}
 
 	fi, err := os.Stat(folder.Path())
-	if v, ok := m.CurrentLocalVersion(id); ok && v > 0 {
```
Comment (Zillode) [inline internal/model/model.go]: Why is this `v > 0` ?

Comment (AudriusButkevicius) [inline internal/model/model.go]: When you initially add a folder, it doesn't have any files, hence version is 0.

**Comment by Zillode** [comment]
Rebased and updated deps, be sure to properly test this after merging please. It works on my machine.
See the following commands:
```
mount -t tmpfs -o size=100m tmpfs /mnt/ramdisk
# Add a folder with the /mnt/ramdisk path
cd /mnt/ramdisk
dd if=/dev/zero of=file.txt count=1022222 bs=1024
```


Should we announce this to events.Default as well?

**Comment by calmh** [comment]
Two more things from mobile (so no line comments for some reason) :) I don't think we handle the `default` annotation in FolderConfiguration so the default needs to be set (and be editable?) in the GUI. And I think DiskUsedPercentage is easier defined as 100-DiskFreePercentage

**Comment by calmh** [comment]
I suggest https://github.com/calmh/syncthing/commit/29646e7e600207d6b837ce89434093925686185e, plus the GUI/config stuff

**Comment by Zillode** [comment]
@st-jenkins run integration tests on linux and windows please

**Comment by calmh** [comment]
rebased & merged
